Blog/variations (#193)
infinum/eightshift-docs: website/blog/2022-09-07-block-variations.md

---
title: Block Variations

> description: Intro to block variations and examples how to use them

Reviewer: ```suggestion
description: Intro to block variations and examples of how to use them
```

slug: block-variations
authors: obradovic
date: 2022-09-07
tags: [eightshift, boilerplate, block, variations]
hide_table_of_contents: false
---

> Let's picture the following scenario. You create a block with many options and want multiple versions of that block available with already set options. That's where variations come in handy!

Reviewer 2: ```suggestion
Let's picture the following scenario: You just created a block with many options and now you want multiple versions of that block available with pre-set options. That's where variations come in handy!
```

<!--truncate-->

## What are variations?

Block variations allow us to override default block attributes. We can select a variation from the block list with all preset options instead of manually setting them.

> Please note that you cannot add new attributes in variations. Only attributes that exist in the parent block can be used.

For example, we have a `Card` block. If we want to use it for something like downloading PDF files, we may not need image or paragraph components. A simple text that describes type of the file, title of the file and a download button are all we need in this case.

Our new card should consist of the following components:
- intro
- heading
- button

## How to register a block variation?

> The process of registering block variations is fairly simple. In order to register a block variation, go to **src/Blocks/variations**, add a new folder, let's call it `card-resource`. Inside that folder, all you need to do is add a `manifest.json` file. It is also recommended to add a `docs` folder in which you can add a readme file and storybook.

Reviewer: ```suggestion
The process of registering block variations is fairly simple. In order to register a block variation, go to **src/Blocks/variations**, add a new folder, and let's call it `card-resource`. Inside that folder, all you need to do is add a `manifest.json` file. It is also recommended to add a `docs` folder in which you can add a readme file and storybook file.
```

Inside your `manifest.json` file, add the attributes to define the new default attributes for this variation. Here is an example:

```json
{
	"$schema": "https://raw.githubusercontent.com/infinum/eightshift-frontend-libs/develop/schemas/variation.json",
	"parentName": "card",
	"name": "card-resource",
	"title": "Card Resource",
	"description" : "Card variation without image and paragraph",
	"icon": {
		"src": "es-card"
	},
	"attributes": {
		"cardCardImageUse": false,
		"cardCardParagraphUse": false,
		"cardCardHeadingSize": "regular",
		"cardCardButtonColor": "blue"
	},
	"scope": [
		"inserter"
	]
}

> 

Reviewer: ```suggestion
```

```

After adding this code, your new block variation should now be visible in the block list. It's that easy!

![Variation of the Card block](/img/blog/card-resource.png)

## Providing inner block data
Other than overriding default attributes with variations, you can do much more with Eightshift Development kit. If you have a block that uses inner blocks, you can even provide inner block data.

The following example is very basic, but it will give you an idea of how to provide inner block data. It can easily be reproduced with our `Carousel` block if you want to try it yourself.

> If you don't have Carousel block in your project yet, you can add it with the following WP-CLI command: `wp boilerplate blocks use-block --name="carousel"`

Once you have the `Carousel` block up and running, create a variation called `Carousel Loop`. Here is the code you can add to the `manifest.json` file for that variation.
```json
{
	"$schema": "https://raw.githubusercontent.com/infinum/eightshift-frontend-libs/develop/schemas/variation.json",
	"parentName": "carousel",
	"name": "carousel-loop",
	"title": "Carousel Loop",

> 	"description" : "Carousel block variation with custom settings",

Reviewer: Can this description be a bit more _descriptive_? 😅

Author: Perhaps 😁 I'll update it 😄

	"icon": {
		"src": "es-card"
	},
	"attributes": {
		"carouselIsLoop": true,
		"carouselShowPagination": true,
		"carouselShowItems": 1
	},
	"innerBlocks": [
		{
			"name": "eightshift-boilerplate/image",
			"attributes": {
				"imageImageFull":true
			}
		},
		{
			"name": "eightshift-boilerplate/image",
			"attributes": {
				"imageImageFull":true
			}
		}
	],
	"scope": [
		"inserter"
	]
}
```

> After adding this code, you should see the `Carousel Loop` variation in your block list. After adding it in the editor, the `Carousel` attributes will be already set to the ones you provided, as well as two images that are added as inner blocks.

Reviewer: How does this carousel differ from the default one? Can you explain this a bit?

Also, the images are empty, maybe we can add some placeholder images? And add maybe 3 or 4 images.

Author: Yeah, I can describe it in a bit more detail 🙂 I didn't think it was too important since it was just a simple example.

I left the images empty on purpose because then the user can immediately add their images, without removing the placeholders first.  But if you think it would be better with placeholders, then I can add some, sure 🙂

Reviewer: I mean, if somebody would just copy/paste the code from the docs in their project it wouldn't hurt to have some images so that they can play around 🤷🏼‍♂️
